python/tutorials/10-block-scaled-matmul.py

```diff
@@ -0,0 +1,335 @@
+        a = tl._experimental_descriptor_load(a_desc, [offs_am, offs_k], [BLOCK_M, BLOCK_K // ELEM_PER_BYTE], dtype)
+        b = tl._experimental_descriptor_load(b_desc, [offs_bn, offs_k], [BLOCK_N, BLOCK_K // ELEM_PER_BYTE], dtype)
+        scale_a = tl.load(a_scale_ptr)
+        scale_b = tl.load(b_scale_ptr)
+        if USE_2D_SCALE_LOAD:
+            scale_a = scale_a.reshape(BLOCK_M // 128, BLOCK_K // VEC_SIZE // 4, 32, 4, 4)
+            scale_b = scale_b.reshape(BLOCK_N // 128, BLOCK_K // VEC_SIZE // 4, 32, 4, 4)
+        scale_a = scale_a.trans(0, 3, 2, 1, 4).reshape(BLOCK_M, BLOCK_K // VEC_SIZE)
+        scale_b = scale_b.trans(0, 3, 2, 1, 4).reshape(BLOCK_N, BLOCK_K // VEC_SIZE)
+        if ELEM_PER_BYTE == 2:
+            accumulator = tl.dot_scaled(a, scale_a, "e2m1", b.T, scale_b, "e2m1", accumulator)
+        else:
+            accumulator = tl.dot_scaled(a, scale_a, "e4m3", b.T, scale_b, "e4m3", accumulator)
+        offs_k += BLOCK_K // ELEM_PER_BYTE
+        a_scale_ptr += (BLOCK_K // VEC_SIZE // 4) * stride_sb
+        b_scale_ptr += (BLOCK_K // VEC_SIZE // 4) * stride_sb
+    tl._experimental_descriptor_store(c_desc, accumulator.to(output_dtype), [offs_am, offs_bn])
+
+
+def block_scaled_matmul(a_desc, a_scale, b_desc, b_scale, dtype_dst, M, N, K, configs):
+    output = torch.empty((M, N), dtype=dtype_dst, device="cuda")
+    if dtype_dst == torch.float32:
+        dtype_dst = 0
+    elif dtype_dst == torch.float16:
+        dtype_dst = 1
+    elif dtype_dst == torch.float8_e4m3fn:
+        dtype_dst = 2
+    else:
+        raise ValueError(f"Unsupported dtype: {dtype_dst}")
+
+    c_desc = TmaDescKernelParam(output.data_ptr(), output.shape, [configs["BLOCK_SIZE_M"], configs["BLOCK_SIZE_N"]],
+                                output.element_size())
+
+    grid = (triton.cdiv(M, configs["BLOCK_SIZE_M"]) * triton.cdiv(N, configs["BLOCK_SIZE_N"]), 1)
+    block_scaled_matmul_kernel[grid](a_desc, a_scale, b_desc, b_scale, c_desc, M, N, K, a_scale.stride(0),
+                                     a_scale.stride(1), a_scale.stride(2), a_scale.stride(3), dtype_dst,
+                                     configs["ELEM_PER_BYTE"], configs["VEC_SIZE"], configs["BLOCK_SIZE_M"],
+                                     configs["BLOCK_SIZE_N"], configs["BLOCK_SIZE_K"], configs["num_stages"],
+                                     USE_2D_SCALE_LOAD=True)
+    return output
+
+
+def initialize_block_scaled(M, N, K, block_scale_type="nvfp4", compute_reference=False):
+    BLOCK_M = 128
+    BLOCK_N = 256
+    BLOCK_K = 256 if "fp4" in block_scale_type else 128
+    VEC_SIZE = 16 if block_scale_type == "nvfp4" else 32
+    assert block_scale_type in ["nvfp4", "mxfp4", "mxfp8"], f"Invalid block scale type: {block_scale_type}"
+    ELEM_PER_BYTE = 2 if "fp4" in block_scale_type else 1
+
+    device = "cuda"
+    a_ref = MXFP4Tensor(size=(M, K), device=device).random()
+    b_ref = MXFP4Tensor(size=(N, K), device=device).random()
```

Comment:
Shall we talk about the col-major requirement on RHS for `fp4 x fp4`? This is the same as fp8 Hopper wgmma.

Reply:
Thank you and good point, I added some discussion on it that should be helpful to users.